__tests__/app.test.js

```diff
@@ -0,0 +1,76 @@
+const db = require('../db/connection')
+const app = require('../app')
+const request = require('supertest')
+const seed = require('../db/seeds/seed')
+const testData = require('../db/data/test-data')
+const endpoints = require('../endpoints.json')
+
+beforeEach(()=>{
+    return seed(testData)
+})
+afterAll(()=>{
+    return db.end()
+})
+describe('GET /*', ()=>{
+    test('should return a 404 status when an attempt to access a non existent endpoint is made', ()=>{
+        return request(app)
+        .get('/non-existing')
+        .expect(404)
+    })
+})
+describe('GET /api/topics', ()=>{
+    test('should return 200 status code', ()=>{
+        return request(app)
+        .get('/api/topics')
+        .expect(200)
+    })
+    test('should return the 3 topics from the table in nc_news_test db', ()=>{
+        return request(app)
+        .get('/api/topics')
+        .then(({body})=>{
+            expect(body.topics).toHaveLength(3)
+        })
+    })
+    test('should return the topics from the table in nc_news_test db with both properties, slug and description', ()=>{
+        return request(app)
+        .get('/api/topics')
+        .then(({body})=>{
+            const topicsArr = body.topics
+            const exampleTopic = {
+                description: 'what books are made of',
+                slug: 'paper'
+              }
+            topicsArr.forEach((topic)=> {
+                for (const key of Object.keys((exampleTopic))){
+                    expect(Object.hasOwn(topic, key))
+                }
+            })
+        })
+    })
+    
+
+})
+describe('GET /api', ()=>{
+    test('should return an object with all valid endpoints and their methods as properties',()=>{
+        return request(app)
+        .get('/api')
+        .then(({body})=>{
+            const appRouteNames = []
+            app._router.stack.forEach((r)=>{
+                if(r.route && r.route.path){
+                    if(r.route.path !== '/*'){ //exclude invalid endpoints
+                        Object.keys(r.route.methods)
+                        .forEach((method) => {
+                            appRouteNames.push(`${method.toUpperCase()} ${r.route.path}`)
+                        })
+                    }
+                } 
+            })
+            let hasAllProperties = true
+            Object.keys(endpoints).forEach((endpoint)=>{
+                if (!appRouteNames.includes(endpoint)) hasAllProperties = false
+            })
+            expect(hasAllProperties).toBe(true)
+        })
+    })
+})
```

Comment:
I can see that you've worked hard at trying to get the appropriate keys from the object here. Try using the file where we know the information is - the endpoints.json. We can compare this to the response to check it's equality.

Reply:
In this test I am inspecting the express app object to see which routes we are listening on and then checking these against the descriptions in  the endpoints.json file. If there is no description for an endpoint this test fails.
So I am in fact using the json file as you suggest... could you clarify if you think this is not a good solution?

Thanks a lot...